providers/multipart/src/main/java/org/jboss/resteasy/plugins/providers/multipart/HeaderFlushedAsyncOutputStream.java

```diff
@@ -0,0 +1,95 @@
+package org.jboss.resteasy.plugins.providers.multipart;
+
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.RuntimeDelegate;
+
+import org.jboss.resteasy.spi.AsyncOutputStream;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
+ * @version $Revision: 1 $
+ */
+public class HeaderFlushedAsyncOutputStream extends AsyncOutputStream {
+   private MultivaluedMap<String, Object> headers;
+   private AsyncOutputStream stream;
+   private boolean headersFlushed = false;
+
+   public HeaderFlushedAsyncOutputStream(final MultivaluedMap<String, Object> headers,
+                                         final AsyncOutputStream delegate) {
+      this.headers = headers;
+      this.stream = delegate;
+   }
+
+   @SuppressWarnings(value = "unchecked")
+   protected CompletionStage<Void> flushHeaders() {
+      CompletionStage<Void> ret = CompletableFuture.completedFuture(null);
+      if (headersFlushed)
+         return ret;
+
+      headersFlushed = true;
+      RuntimeDelegate delegate = RuntimeDelegate.getInstance();
+
+      for (String key : headers.keySet()) {
+         List<Object> objs = headers.get(key);
+         for (Object obj : objs) {
+            String value;
+            RuntimeDelegate.HeaderDelegate headerDelegate = delegate
+                  .createHeaderDelegate(obj.getClass());
+            if (headerDelegate != null) {
+               value = headerDelegate.toString(obj);
+            } else {
+               value = obj.toString();
+            }
+            ret = stream.asyncWrite(key.getBytes())
+                    .thenCompose(v -> stream.asyncWrite(": ".getBytes()))
+                    .thenCompose(v -> stream.asyncWrite(value.getBytes()))
+                    .thenCompose(v -> stream.asyncWrite("\r\n".getBytes()));
```

Comment:
We can't without chaining promises anyway in many cases, so there will still be underlying costs.

Actually it can be cheaper but only in the case we buffer and flush in chunks, which is the case for vertx/netty backends but not quarkus/servlet, so IMO it's not worth it for us ATM.

Reply:
Servlet always buffers, the content is copied into a direct buffer internally, and it does not actually flush to the wire until the buffer is full.

Reply:
Servlet buffers, but internally, so I don't see it and still have to call `write(byte[])` and wire the notifications, so I still pay the cost of wiring them.